proposals/low-level-struct-improvements.md

```diff
@@ -0,0 +1,972 @@
+account the lifetime of `this` (because it can't return a `ref` to internal
+state). This means that `ref` returns from a `struct` can return outside the 
+enclosing method scope except in cases where there are `ref` parameters or a 
+`ref struct` which is not *safe-to-escape* outside the enclosing method scope. 
+Hence the solution here is not as easy as allowing `ref` return of fields in 
+non-interface methods.
+
+To remove this friction the language will provide the attribute `[RefEscapes]`.
+When this attribute is applied to an instance method, instance property or 
+instance accessor of a `struct` or `ref struct` then the `this` parameter will
+be considered *ref-safe-to-escape* outside the enclosing method.
+
+This allows for greater flexibility in `struct` definitions as they can begin 
+returning `ref` to their fields. That allows for types like `FrugalList<T>`:
+
+```cs
+struct FrugalList<T>
+{
+    private T _item0;
+    private T _item1;
+    private T _item2;
+
+    public int Count = 3;
+
+    public ref T this[int index]
+    {
+        [RefEscapes]
+        get
+        {
+            switch (index)
+            {
+                case 0: return ref _item1;
+                case 1: return ref _item2;
+                case 1: return ref _item3;
+                default: throw null;
+            }
+        }
+    }
+}
+```
+
+This will naturally, by the existing rules in the span safety spec, allow
+for returning transitive fields in addition to direct fields.
+
+```cs
+struct ListWithDefault<T>
+{
+    private FrugalList<T> _list;
+    private T _default;
+
+    public ref T this[int index]
+    {
+        [RefEscapes]
+        get
+        {
+            if (index >= _list.Count)
+            {
+                return ref _default;
+            }
+
+            return ref _list[index];
+        }
+    }
+}
+```
+
+Members which contain the `[RefEscapes]` attribute cannot be used to implement 
+interface members. This would hide the lifetime nature of the member at 
+the `interface` call site and would lead to incorrect lifetime calculations.
+
+To account for this change the "Parameters" section of the span safety document
+will be updated to include the following:
+
+- If the parameter is the `this` parameter of a `struct` type, it is 
+*ref-safe-to-escape* to the top scope of the enclosing method unless the 
+method is annotated with `[RefEscapes]` in which case it is *ref-safe-to-escape*
+outside the enclosing method.
+
+Misc Notes:
+- A member marked as `[RefEscapes]` can not implement an `interface` method.
+- The `RefEscapesAttribute` will be defined in the 
+`System.Runtime.CompilerServices` namespace.
+
+### Safe fixed size buffers
+The language will relax the restrictions on fixed sized arrays such that the 
+can be declared in safe code and the element type can be managed or unmanaged. 
+This will make types like the following legal:
+
+```cs
+internal struct CharBuffer
+{
+    internal fixed char Data[128];
+}
+```
```

Comment:
Is there a story if I want to use a buffer as a local?  Or I will have to wrap in a struct?

I have seen some use of `stackalloc` in corefx where a fixed sized buffer local would fit better. 
Typically small buffers of fixed size, used in a loop to parse something. 
stackallocs are not very friendly to use in loops.

Reply:
(Example: https://github.com/dotnet/runtime/pull/34149)

Reply:
I had not thought about this case but I can see the value that it would provide. Will think on this a bit and see if it's easy enough to make work.

Reply:
Added this to the "Open Issues" section of the design so we can discuss if the complexity is worth it in LDM. This section actually hasn't been discussed in LDM yet so this can just be a part of that discussion.